aws-lc-rs/src/agreement.rs

```diff
@@ -523,32 +565,81 @@ where
     F: FnOnce(&[u8]) -> Result<R, E>,
 {
     let expected_alg = my_private_key.algorithm();
-    let expected_pub_key_len = expected_alg.id.pub_key_len();
     let expected_nid = expected_alg.id.nid();
 
     if peer_public_key.alg != expected_alg {
         return Err(error_value);
     }
     let peer_pub_bytes = peer_public_key.bytes.as_ref();
-    if peer_pub_bytes.len() != expected_pub_key_len {
-        return Err(error_value);
+    if peer_public_key.encoding == &OCTET_STRING {
+        let expected_pub_key_len = expected_alg.id.pub_key_len();
+        let peer_pub_bytes_len = peer_pub_bytes.len();
+        if peer_pub_bytes_len != expected_pub_key_len {
+            return Err(error_value);
+        }
     }
 
     let mut buffer = [0u8; MAX_AGREEMENT_SECRET_LEN];
 
     let secret: &[u8] = match &my_private_key.inner_key {
-        KeyInner::X25519(priv_key) => {
-            x25519_diffie_hellman(&mut buffer, priv_key, peer_pub_bytes).or(Err(error_value))?
-        }
-        KeyInner::ECDH_P256(priv_key)
-        | KeyInner::ECDH_P384(priv_key)
-        | KeyInner::ECDH_P521(priv_key) => {
-            ec_key_ecdh(&mut buffer, priv_key, peer_pub_bytes, expected_nid).or(Err(error_value))?
+        KeyInner::X25519(priv_key, ..) => match peer_public_key.encoding.id {
+            EncodingID::OctetString => {
+                x25519_diffie_hellman(&mut buffer, priv_key, peer_pub_bytes)
+                    .map_err(|()| error_value)?;
+                &buffer[0..X25519_SHARED_KEY_LEN]
+            }
+            EncodingID::X509 => {
+                x25519_dh_with_x509_peer_pubkey(
+                    peer_pub_bytes,
+                    &mut buffer,
+                    priv_key,
+                    error_value,
+                )?;
+                &buffer[0..X25519_SHARED_KEY_LEN]
+            }
+        },
+        KeyInner::ECDH_P256(ec_key) | KeyInner::ECDH_P384(ec_key) | KeyInner::ECDH_P521(ec_key) => {
+            let pub_key_bytes = peer_public_key.bytes.as_ref();
+            ec_key_ecdh(
+                &mut buffer,
+                ec_key,
+                pub_key_bytes,
+                peer_public_key.encoding,
+                expected_nid,
+            )
+            .or(Err(error_value))?
         }
     };
     kdf(secret)
 }
 
+#[inline]
+fn x25519_dh_with_x509_peer_pubkey<E: Copy>(
+    peer_x509_pubkey: &[u8],
+    shared_secret: &mut [u8; MAX_AGREEMENT_SECRET_LEN],
+    priv_key: &LcPtr<EVP_PKEY>,
+    error_value: E,
+) -> Result<(), E> {
+    let len = match c_long::try_from(peer_x509_pubkey.len()) {
+        Ok(len) => len,
+        Err(_) => return Err(error_value),
+    };
+    let x509_pubkey = LcPtr::new(unsafe {
+        d2i_X509_PUBKEY(
+            null_mut(),
+            &mut peer_x509_pubkey.as_ptr() as *const *const u8 as *mut *const c_uchar,
```

Comment:
```suggestion
            &mut peer_x509_pubkey.as_ptr() as *mut *const u8 as *mut *const c_uchar,
```

nit: this is likely not important at runtime but helps avoid future lints etc.

Reply:
Will change.